terminus-ui/popover/src/popover-trigger.directive.ts

```diff
@@ -0,0 +1,302 @@
+import {
+  AfterContentChecked,
+  ChangeDetectorRef,
+  Directive,
+  ElementRef,
+  EventEmitter,
+  HostListener,
+  Input,
+  NgZone,
+  OnChanges,
+  OnDestroy,
+  OnInit,
+  Output,
+  SimpleChange,
+  ViewContainerRef,
+} from '@angular/core';
+import {
+  isUnset,
+  KEYS,
+  TsDocumentService,
+  untilComponentDestroyed,
+} from '@terminus/ngx-tools';
+import {
+  ControlValueAccessorProviderFactory,
+  TsUILibraryError,
+} from '@terminus/ui/utilities';
+import {
+  fromEvent,
+  Observable,
+} from 'rxjs';
+import { filter } from 'rxjs/operators';
+import {
+  PopoverOptions,
+  Triggers,
+  TsPopoverPosition,
+  tsPopoverPositions,
+  TsPopoverPositions,
+} from './popover-options';
+import { TsPopoverComponent } from './popover.component';
+
+// Unique ID for each instance
+let nextUniqueId = 0;
+
+/**
+ * A directive that adds popover trigger functionality
+ *
+ * @example
+ * <button
+ *              tsPopoverTrigger="popper1"
+ *              [position]="position"
+ *              [popover]="popper1"
+ *              [defaultOpened]="defaultOpened"
+ *              (popoverOnShown)="popoverOnShown()"
+ *              (popoverOnHidden)="popoverOnHidden()"
+ * />
+ */
+@Directive({
+  selector: '[tsPopoverTrigger]',
+  host: { class: 'ts-popover-trigger' },
+  providers: [
+    ControlValueAccessorProviderFactory<TsPopoverTriggerDirective>(TsPopoverTriggerDirective),
+  ],
+  exportAs: 'tsPopoverTrigger',
+})
+export class TsPopoverTriggerDirective implements OnInit, OnDestroy, OnChanges, AfterContentChecked {
+
+  /**
+   * Store a reference to the document object
+   */
+  private document: Document;
+
+  /**
+   * When no id provided, it falls back to uid
+   */
+  public readonly uid = `ts-popover-trigger-${nextUniqueId++}`;
+
+  /**
+   * Some default options for popper. For now we only support click, so it specifies click here. But could support more if needed.
+   */
+  public static defaultOptions: PopoverOptions = {
+    trigger: Triggers.CLICK,
+    ariaRole: 'popover',
+    placement: TsPopoverPositions.Bottom,
+  };
+
+  /**
+   * Whether current popover is open or not.
+   */
+  public isOpen = false;
```

Comment:
I see us setting this flag, but is it ever used anywhere? Just wondering if we can use a better method to know if the popper is open or not. It seems that we should be able to get this information from either Popper.js or our popper component (since it knows when it's opened to fire the emitters). Just hoping for a less manual way to know since currently we set the flag without actually knowing if the popper was opened.

Reply:
This flag is only used internally by trigger... mainly so toggle would know to toggle it to open or hide. I had considered  making it subscribing to the event emitter, but thought it's an overkill since it's not trying to keep track of status between component/directive. popover only opens on load when `defaultOpened` set to `true`, where `ngAfterContentCheck` will check. I think it works well for our current case, but if we consider supporting more complex case down the road, I agree that subscribing to an event will make more sense.